src/openfermion/ops/_symbolic_operator.py

```diff
@@ -0,0 +1,520 @@
+                # it ends
+                if not factor[0].isdigit():
+                    raise ValueError('Invalid factor {}.'.format(factor))
+                index_end = 1
+                while (index_end <= len(factor) - 1 and
+                        factor[index_end].isdigit()):
+                    index_end += 1
+
+                index = int(factor[:index_end])
+                action_string = factor[index_end:]
+            # Check that the index is valid
+            if index < 0:
+                raise ValueError('Invalid index in factor {}. '
+                                 'The index should be a non-negative '
+                                 'integer.'.format(factor))
+            # Convert the action string to an action
+            if action_string in self.action_strings:
+                action = self.actions[self.action_strings.index(action_string)]
+            else:
+                raise ValueError('Invalid action in factor {}. '
+                                 'Valid actions are: {}'.format(
+                                     factor, self.action_strings))
+
+            # Add the factor to the list as a tuple
+            processed_term.append((index, action))
+
+        # If factors with different indices commute, sort the factors
+        # by index
+        if self.different_indices_commute:
+            processed_term = sorted(processed_term,
+                                    key=lambda factor: factor[0])
+
+        # Return a tuple
+        return tuple(processed_term)
+
+    @classmethod
+    def zero(cls):
+        """
+        Returns:
+            additive_identity (SymbolicOperator):
+                A symbolic operator o with the property that o+x = x+o = x for
+                all operators x of the same class.
+        """
+        return cls(term=None)
+
+    @classmethod
+    def identity(cls):
+        """
+        Returns:
+            multiplicative_identity (SymbolicOperator):
+                A symbolic operator u with the property that u*x = x*u = x for
+                all operators x of the same class.
+        """
+        return cls(term=())
+
+    def __str__(self):
+        """Return an easy-to-read string representation."""
+        if not self.terms:
+            return '0'
+        string_rep = ''
+        for term in self.terms:
+            tmp_string = '{} ['.format(self.terms[term])
+            for factor in term:
+                index, action = factor
+                action_string = self.action_strings[self.actions.index(action)]
+                if self.action_before_index:
+                    tmp_string += '{}{} '.format(action_string, index)
+                else:
+                    tmp_string += '{}{} '.format(index, action_string)
+            string_rep += '{}] +\n'.format(tmp_string.strip())
+        return string_rep[:-3]
+
+    def __repr__(self):
+        return str(self)
+
+    def __imul__(self, multiplier):
```

Comment:
So is the idea that for QubitOperator we will overwrite this class? Because in that case there is a more optimized way to multiply which is based on knowledge of the fact that the QubitOperator terms are sorted by tensor factor. I think that is fine, I am just trying to understand.

Reply:
Yes, the idea is that QubitOperator will overwrite this method using its existing implementation.